### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#617] [Ideas]
# Infinity data source plugin roadmap

## potential candidates for upcoming versions

- [ ] token refresh authentication
- [ ] prom/loki agent
- [ ] local time-shift
- [ ] Join results from multiple APIs

## Version 2.0.0

- [x] https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/601
- [x] https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/614
- [x] move backend experimental features out of beta

## Version 1.0.0

- [x] OAuth support
- [x] Basic/experimental backend support for common JSON formats
- [x] Support for Grafana unified Alerting
- [x] Support for Grafana Server Side Expressions
- [x] Support for Grafana Recorded queries

## Version 0.10.0

- [x] ~Client side query caching~
- [x] ~In JSON, Add support to access fields outside array~ ( use UQL instead )
- [x] Dynamic interval variable
- [x] Migrate string based legacy variables

## Version 0.9.0

- [x] Support for [GROQ query language](https://www.sanity.io/docs/groq)
- [x] Filter data with UQL
- [x] #196 
- [x] Show corresponding curl command in Query inspector
- [x] Forward OAuth identity support

**Comment by ebachman24** [comment]
Being able to handle paginated results would be incredible.  Great work!  looking forward to it.

**Comment by gdiana** [comment]
The backend stuff is great well done and many thanks.

I saw a GRAFANA post that looked at being able to update the backend csv file to import added new data. Can this be done or is it already available under the GRAFAN update source feature?

**Comment by kayjayeff** [comment]
Personally, I would really like to see the "refresh token" ability put into the next version.  This is something that does not seem to be out there in any API plugin.  I have run into multiple data sources that require an initial token and then the refresh.  Infinity is a great and versatile plugin and adding that functionality would make it even moreso.

**Reply by Moktad** [reply]
Yeah same here but I'm trying to find a workaround to somehow refresh the token 
Please let me know if you found something about it.

**Reply by KevinDW-Fluxys** [reply]
+1

**Comment by jHumeLPL** [comment]
> * [ ]  token refresh authentication
> * [ ]  local time-shift
> * [ ]  Join results from multiple APIs

Assuming these 3 features work the way I hope, I am really excited for all of them, especially the token refresh and the local time-shift.

I assume the **token refresh** is to make querying API's that require getting a new token before querying them easier.

I assume the **local time-shift** is to allow changing timestamp values coming in from the API from one time zone to another.  In our case, there's an API we query and all of the timestamps it returns are in Eastern, but of course Grafana requires UTC time.  We have found a way to "fix" it using several different Grafana Transforms in the Query Editor, but it is a huge pain and being able to do it at the data source layer, before hitting the Grafana Visualization/Variables would be far better.

And wow, the more I think about the **Join results from multiple APIs** idea, the more I like that as well...  I can think of several uses of that for us.

Keep up the great work!

**Comment by giosueDelgado** [comment]
Hello there are some update?
